### PR TITLE
Migrating changes from special release 0.4 to master

### PR DESCRIPTION
**Pull request by troels-im**

### CODE REVIEW

**Review by pbukva — DISMISSED** [review]
LGTM

**Review by pbukva — DISMISSED** [review]
LGTM

File: libs/ledger/src/chaincode/smart_contract.cpp [hunk 1]
```diff
@@ -278,6 +279,45 @@ void AddAddressToParameterPack(vm::VM *vm, vm::ParameterPack &pack, variant::Var
   pack.Add(vm_address);
 }
 
+/**
+ * Extract an address from a msgpack::object
```
Comment (ejfitzgerald) [inline libs/ledger/src/chaincode/smart_contract.cpp]: Documentation doesn't seem to match function 😁

File: libs/meta/include/meta/has_index.hpp [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+
+namespace fetch {
+namespace meta {
+
+namespace detail {
+template <class T, class Index>
+struct HasIndexImpl
+{
+  template <class T1,
+            class IndexDeduced = Index,  // <--- here
```
Comment (ejfitzgerald) [inline libs/meta/include/meta/has_index.hpp]: `<---` still needed?

File: libs/vm/include/vm/map.hpp [hunk 3]
```diff
@@ -179,17 +179,110 @@ struct Map : public IMap
 
   bool SerializeTo(ByteArrayBuffer &buffer) override
   {
-    buffer << map;
+    //    TypeInfo const &type_info     = vm_->GetTypeInfo(GetTypeId());
```
Comment (ejfitzgerald) [inline libs/vm/include/vm/map.hpp]: Not needed anymore

**Review by ejfitzgerald — CHANGES_REQUESTED** [review]
I mostly get the flow of changes here. Quite a lot of changes. I note that there are quite a few trailing TODO issues can we update them to:

`TODO(issue XXX): Foo bar baz`

So that we can track them.

**Review by ejfitzgerald — APPROVED** [review]
🚢 it!

**Pull request merged**